docs/DesignDocs/Protected_Code_Loader_Proposal.md

```diff
@@ -0,0 +1,161 @@
+Open Enclave Protected Code Loader
+====
+
+Enabling Enclave Code Confidentiality in Open Enclave.
+
+# Motivation
+Current Open Enclave SDK provides code integrity and data confidentiality at
+run-time but not enclave binary confidentiality on a disk. The enclave binary
+is in plaintext and can be reverse engineered to reveal code logic and
+secret data embedded in the enclave image. 
+
+The Open Enclave Protected Code Loader shall enable providing confidentiality 
+and integrity to the IP sections in enclave image, i.e. , the ELF file.
+
+# User Experience
+## Enclave Build-time:
+### Linking pcl lib to the enclave 
+On building the enclave, a static lib "liboepcl.a" should be linked to the 
+enclave image with the compilation option "-Wl, whole-archive"
+### Encrypt the enclave image
+An encryption tool is provided as the pre-step to enclave signing process:
+
+```bash
+oeencrypt -i my_enclave -o my_enclave.enc -k keyfile
+```
+
+## Enclave Load-time:
+On Enclave Loading, if the image is encrypted, it need to be decrypted 
+inside enclave loader, this step is done by OE SDK runtime and invisible 
+to user. 
+
+# Specification
+
+## Protected Code Loader Software Work Flow in Open Enclave
+### Build time encryption 
+    A seperate tool "oeencrypt" is provided at enclave build time to encrypt 
+	the enclave image right before enclave signing process.  
+### Load time decryption
+    On enclave initliazation operation(EINIT operation), right before relocation, 
+    the encrypted enclave image need to be decrypted and than perform the relocation
+    operation.
+
+## Encryption Algorithm in Protected Code Loader
+Using openssl AES-256-GCM as the encryption/decryption algorithm.
```

Comment:
Can we support crypto-agility?

Reply:
I think AES GCM algorithm is the most popular encrypt/decrypt algorithms used today. And for users wants to use this feature, they don't need to do more work on this encryption/decryption part. Also as mentioned in Cedric's PR 3652, the GCM is faster than other block cipher algorithms.

Reply:
I'm not an expert so just asking, but https://tools.ietf.org/html/rfc7696#section-3.1 says:

>    Sometimes more than one mandatory-to-implement algorithm is needed to
>    increase the likelihood of interoperability among a diverse
>    population.  For example, authenticated encryption is provided by
>    AES-CCM [RFC3610] and AES-GCM [GCM].  Both of these algorithms are
>    considered to be secure.  AES-CCM is available in hardware used by
>    many small devices, and AES-GCM is parallelizable and well suited to
>    high-speed devices.  Therefore, an application needing authenticated
>    encryption might specify one of these algorithms or both of these
>    algorithms, depending on the population.

Hence my question, since OE SDK is not just for high-speed devices like cloud servers.
If a crypto expert tells me that AES-256-GCM is sufficient for IoT, and will suffice for many years to come, then it's fine with me, but I have to ask.

Reply:
Just as an aside: it is likely that AES-256-GCM will remain in use for many years; the CAESAR competition I think only announced the results early last year. If you compare that to when SHA-3 was decided (back in 2012) and the ramp for general adoption of it, you might have a sense for how long it might take the new authenticated encryption standard to become pervasive.

An interesting thing to note is that CAESAR breaks the choice of algorithms down between resource-constrained and high-performance app categories, so in the future, it may be that IoT may choose a different algorithm than something running on server.